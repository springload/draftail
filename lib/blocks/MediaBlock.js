import React, { Component } from 'react';

import { ENTITY_TYPE } from '../api/constants';

import ImageBlock from '../blocks/ImageBlock';
import EmbedBlock from '../blocks/EmbedBlock';
import NullBlock from '../blocks/NullBlock';

const BLOCK_MAP = {};

BLOCK_MAP[ENTITY_TYPE.IMAGE] = ImageBlock;
BLOCK_MAP[ENTITY_TYPE.EMBED] = EmbedBlock;

/**
 * Editor block to display media and edit content.
 */
class MediaBlock extends Component {
    constructor(props) {
        super(props);

        this.state = {
            isActive: false,
        };

        this.toggleBlock = this.toggleBlock.bind(this);
        this.closeBlock = this.toggleBlock.bind(this, false);
        this.onClick = this.onClick.bind(this);
        this.onSave = this.onSave.bind(this);
    }

    toggleBlock(isActive = !this.state.isActive) {
        const { blockProps } = this.props;

        this.setState({
            isActive: isActive,
        });

        if (isActive) {
            blockProps.lockEditor();
        } else {
            blockProps.unlockEditor();
        }
    }

    onClick() {
        this.toggleBlock();
    }

    onSave(nextData) {
        const { block, contentState } = this.props;

        // This will update in place
        contentState.mergeEntityData(block.getEntityAt(0), nextData);
        this.closeBlock();
    }

    render() {
        const { blockProps } = this.props;
        const { isActive } = this.state;
        const Block = BLOCK_MAP[blockProps.entity.type] || NullBlock;

        return (
            <div className={`RichEditor-media${isActive ? ' RichEditor-media--open' : ''}`}>
                <Block
                    entity={blockProps.entity}
                    entityConfig={blockProps.entityConfig}
                    isActive={isActive}
                    onClick={this.onClick}
                    onCancel={this.closeBlock}
                    onSave={this.onSave}
                />
            </div>
        );
    }
}

MediaBlock.propTypes = {
    block: React.PropTypes.object.isRequired,
    contentState: React.PropTypes.object.isRequired,
    blockProps: React.PropTypes.object.isRequired,
};

export default MediaBlock;
