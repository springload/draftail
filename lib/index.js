/**
 * Draftail's main API entry point. Exposes all of the modules people
 * will need to create their own editor instances from Draftail.
 */
export { default as DraftailEditor } from "./components/DraftailEditor";
export { default as Icon } from "./components/Icon";
export { default as ToolbarButton } from "./components/ToolbarButton";
export { default as DraftUtils } from "./api/DraftUtils";
export { BLOCK_TYPE, ENTITY_TYPE, INLINE_STYLE } from "./api/constants";
