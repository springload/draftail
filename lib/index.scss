@import "./api/constants";

@import "./components/Icon";
@import "./components/DraftailEditor";
@import "./components/Toolbar";
@import "./components/ToolbarGroup";
@import "./components/ToolbarButton";
@import "./components/ToolbarTooltip";

@import "./blocks/DividerBlock";
