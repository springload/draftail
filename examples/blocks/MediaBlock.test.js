import React from 'react';
import { shallow } from 'enzyme';
import { convertFromRaw, EditorState } from 'draft-js';

import { ENTITY_TYPE } from '../../lib';

import MediaBlock from '../blocks/MediaBlock';

const contentState = convertFromRaw({
    entityMap: {
        '0': {
            type: ENTITY_TYPE.IMAGE,
            mutability: 'IMMUTABLE',
            data: {
                src:
                    'https://springload.github.io/draftail/static/example-lowres-image.jpg',
            },
        },
    },
    blocks: [
        {
            key: 'b3kdk',
            text: ' ',
            type: 'atomic',
            depth: 0,
            inlineStyleRanges: [],
            entityRanges: [
                {
                    offset: 0,
                    length: 1,
                    key: 0,
                },
            ],
            data: {},
        },
    ],
});

const mockProps = {
    block: contentState.getFirstBlock(),
    blockProps: {
        editorState: EditorState.createWithContent(contentState),
        entity: {
            getType: () => ENTITY_TYPE.IMAGE,
        },
        entityConfig: {
            type: ENTITY_TYPE.IMAGE,
        },
        onChange: jest.fn(),
        lockEditor: jest.fn(),
        unlockEditor: jest.fn(),
    },
    contentState,
};

describe('MediaBlock', () => {
    it('basic', () => {
        expect(shallow(<MediaBlock {...mockProps} />)).toMatchSnapshot();
    });

    it('image', () => {
        expect(
            shallow(
                <MediaBlock
                    {...mockProps}
                    blockProps={Object.assign({}, mockProps.blockProps, {
                        entity: {
                            getType: () => ENTITY_TYPE.IMAGE,
                        },
                    })}
                />,
            ),
        ).toMatchSnapshot();
    });

    it('embed', () => {
        expect(
            shallow(
                <MediaBlock
                    {...mockProps}
                    blockProps={Object.assign({}, mockProps.blockProps, {
                        entity: {
                            getType: () => ENTITY_TYPE.EMBED,
                        },
                        entityConfig: {
                            type: ENTITY_TYPE.EMBED,
                        },
                    })}
                />,
            ),
        ).toMatchSnapshot();
    });

    it('#isActive', () => {
        expect(
            shallow(<MediaBlock {...mockProps} />).setState({
                isActive: true,
            }),
        ).toMatchSnapshot();
    });

    it('#blockProps.lockEditor #blockProps.unlockEditor', () => {
        const lockEditor = jest.fn();
        const unlockEditor = jest.fn();

        const wrapper = shallow(
            <MediaBlock
                {...mockProps}
                blockProps={Object.assign({}, mockProps.blockProps, {
                    lockEditor,
                    unlockEditor,
                })}
            />,
        );

        wrapper.instance().onClick();
        expect(lockEditor).toHaveBeenCalledTimes(1);
        wrapper.instance().onClick();
        expect(unlockEditor).toHaveBeenCalledTimes(1);
    });

    it('#blockProps.onChange', () => {
        const onChange = jest.fn();

        const wrapper = shallow(
            <MediaBlock
                {...mockProps}
                blockProps={Object.assign({}, mockProps.blockProps, {
                    onChange,
                })}
            />,
        );

        wrapper.instance().onSave({
            value: 'test',
        });
        expect(onChange).toHaveBeenCalledTimes(1);
    });
});
