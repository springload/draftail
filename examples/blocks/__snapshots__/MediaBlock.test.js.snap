// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MediaBlock renders 1`] = `
<button
  aria-label="Test: test"
  className="MediaBlock"
  onMouseUp={[Function]}
  tabIndex={-1}
  type="button"
>
  <span
    aria-hidden={true}
    className="MediaBlock__icon-wrapper"
  >
    <Icon
      className="MediaBlock__icon"
      icon="#icon-test"
      title={null}
    />
  </span>
  <img
    alt=""
    className="MediaBlock__img"
    src="example.png"
    width="256"
  />
</button>
`;

exports[`MediaBlock tooltip opens 1`] = `
<div>
  <div>
    <div
      class="Tooltip Tooltip--top-left"
      role="tooltip"
      style="top: 0px; left: 0px;"
    >
      <div
        style="max-width: 300px;"
      >
        Test
      </div>
    </div>
  </div>
</div>
`;
